internal/authorization/serviceaccount_authenticator.go

```diff
@@ -0,0 +1,178 @@
+	client         client.Interface
+	next           http.RoundTripper
+	clientID       string
+	clientSecret   string
+	keyID          string
+	privateKeyData string
+}
+
+func (saa *serviceAccountAuthenticator) RoundTrip(req *http.Request) (*http.Response, error) {
+	return roundTrip(req, saa.next, saa.AccessToken)
+}
+
+func (saa *serviceAccountAuthenticator) AccessToken() (*oauth2.Token, error) {
+	saa.mutex.Lock()
+	defer saa.mutex.Unlock()
+
+	jwt := saa.userAuth.ReadJWTToken()
+
+	if jwt.Valid() {
+		return jwt, nil
+	}
+
+	return saa.basicAuth()
+}
+
+func (saa *serviceAccountAuthenticator) basicAuth() (*oauth2.Token, error) {
+	var jwt *oauth2.Token
+	var err error
+	switch {
+	case len(saa.clientID) > 0 && len(saa.clientSecret) > 0:
+		jwt, err = getClientCredentialsToken(context.Background(), saa.client, saa.clientID, saa.clientSecret)
+	case len(saa.clientID) > 0 && len(saa.keyID) > 0 && len(saa.privateKeyData) > 0:
+		var key *rsa.PrivateKey
+		key, err = rsaKeyFromBase64(saa.privateKeyData)
+		if err != nil {
+			break
+		}
+		jwt, err = getJWTToken(context.Background(), saa.client, saa.keyID, saa.clientID, key)
+	default:
+		err = fmt.Errorf("inconsistent auth configuration")
+	}
+
+	if jwt != nil {
+		saa.userAuth.WriteJWTToken(jwt)
+	}
+
+	return jwt, err
+}
+
+func getClientCredentialsToken(ctx context.Context, apiClient client.Interface, clientID, clientSecret string) (*oauth2.Token, error) {
+	req := apiClient.Post().APIPath(serviceAccountAuthEndpoint)
+	endpoint := req.URL()
+
+	config := clientcredentials.Config{
+		ClientID:     clientID,
+		ClientSecret: clientSecret,
+		TokenURL:     endpoint.String(),
+		AuthStyle:    oauth2.AuthStyleInHeader,
+	}
+
+	tokenContext := context.WithValue(ctx, oauth2.HTTPClient, apiClient.HTTPClient())
+	return config.Token(tokenContext)
+}
+
+func getJWTToken(ctx context.Context, apiClient client.Interface, keyID, clientID string, key *rsa.PrivateKey) (*oauth2.Token, error) {
+	jwsHeader := &jws.Header{
+		Typ:       "JWT",
+		Algorithm: "RS256",
+		KeyID:     keyID,
+	}
+
+	jwsClaim := &jws.ClaimSet{
+		Iss: clientID,
+		Sub: clientID,
+		Aud: "console-client-credentials",
+		PrivateClaims: map[string]interface{}{
+			"jti": uuid.New(),
+		},
+	}
+
+	signedJWS, err := jws.Encode(jwsHeader, jwsClaim, key)
+	if err != nil {
+		return nil, err
+	}
+
+	values := make(url.Values)
+	values.Set("grant_type", "client_credentials")
+	values.Set("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer")
+	values.Set("client_assertion", signedJWS)
+	values.Set("client_id", clientID)
+	values.Set("token_endpoint_auth_method", "private_key_jwt")
+	response, err := apiClient.
+		Post().
+		APIPath(serviceAccountAuthEndpoint).
+		SetHeader("Content-Type", "application/x-www-form-urlencoded").
+		Body([]byte(values.Encode())).
+		Do(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.Error() != nil {
+		return nil, response.Error()
+	}
+
+	type tempToken struct {
```

Comment:
It could be better to define the struct outside the function. It's more a tempTokenResponse struct?

Reply:
I don't need it anywhere else than in that function, and I need it only because oauth2 don't expose the intermediate representation of the Token struct